pallets/randomness/src/types.rs

```diff
@@ -0,0 +1,306 @@
+		let amt_unreserved = amt_to_unreserve.saturating_sub(amt_not_unreserved);
+		let refundable_amount = if amt_unreserved < self.fee {
+			// EDGE CASE: amount unreserved is not equal to deposit + fee
+			// If the amount unreserved is less than the `fee`, we use the entire amount unreserved
+			// to refund caller of fulfill. The `deposit` acts as a safety margin for the refund.
+			amt_unreserved
+		} else {
+			self.fee
+		};
+		if let Some(excess) = refundable_amount.checked_sub(&cost_of_execution) {
+			// refund cost_of_execution to caller of `fulfill`
+			T::ReserveCurrency::transfer(
+				&contract_address,
+				&T::AddressMapping::into_account_id(caller.clone()),
+				cost_of_execution,
+				KeepAlive,
+			)
+			.expect("just unreserved deposit + fee => cost_of_execution must be transferrable");
+			// refund excess to refund address
+			T::ReserveCurrency::transfer(
+				&contract_address,
+				&T::AddressMapping::into_account_id(self.refund_address),
+				excess,
+				KeepAlive,
+			)
+			.expect("just unreserved deposit + fee => excess must be transferrable");
+		} // TODO: else should log warning or emit event that no refund happened?
+	}
+}
+
+#[derive(PartialEq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct RequestState<T: Config> {
+	/// Underlying request
+	pub request: Request<T>,
+	/// Deposit taken for making request (stored in case config changes)
+	pub deposit: BalanceOf<T>,
+	/// All requests expire `T::ExpirationDelay` blocks after they are made
+	pub expires: T::BlockNumber,
+}
+
+#[derive(PartialEq, Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+/// Data required to make the subcallback and finish fulfilling the request
+pub struct FulfillArgs<T: Config> {
+	/// Original request
+	pub request: Request<T>,
+	/// Deposit for request
+	pub deposit: BalanceOf<T>,
+	/// Randomness
+	pub randomness: [u8; 32],
+}
+
+impl<T: Config> RequestState<T> {
+	pub(crate) fn new(request: Request<T>) -> RequestState<T> {
+		let expires =
+			frame_system::Pallet::<T>::block_number().saturating_add(T::ExpirationDelay::get());
+		// TODO: check that request.info.when is before `expires` (how to do it with EpochIndex)
+		RequestState {
+			request,
+			deposit: T::Deposit::get(),
+			expires,
+		}
+	}
+	/// Returns Ok(FulfillArgs) if successful
+	/// This should be called before the callback
+	pub fn prepare_fulfill(&self) -> Result<FulfillArgs<T>, DispatchError> {
+		// get the randomness corresponding to the request
+		let randomness: T::Hash = self.request.get_randomness()?;
+		// compute random output using salt
+		let randomness = Pallet::<T>::concat_and_hash(randomness, self.request.salt);
+		// No event emitted until fulfillment is complete
+		Ok(FulfillArgs {
+			request: self.request.clone(),
+			deposit: self.deposit,
+			randomness,
+		})
+	}
+	pub fn increase_fee(
```

Comment:
The name might suggest that a higher fee would be paid rather than specifying a higher potential fee. Maybe a different name (`increase_max_fee` or similar) or just simply some docs would help.